pkg/util/deployment/deployment.go

```diff
@@ -144,6 +136,158 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
 	return nil, nil
 }
 
+// rsAndPodsWithHashKeySynced returns a list of rs the deployment targets, with pod-template-hash information synced.
+func rsAndPodsWithHashKeySynced(deployment extensions.Deployment, c clientset.Interface, getRSList rsListFunc, getPodList podListFunc) ([]extensions.ReplicaSet, *api.PodList, error) {
+	namespace := deployment.Namespace
+	selector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return nil, nil, err
+	}
+	options := api.ListOptions{LabelSelector: selector}
+	rsList, err := getRSList(namespace, options)
+	if err != nil {
+		return nil, nil, err
+	}
+	syncedRSList := []extensions.ReplicaSet{}
+	for _, rs := range rsList {
+		// Add pod-template-hash information if it's not in the RS.
+		// Otherwise, new RS produced by Deployment will overlap we pre-existing ones
+		// that aren't constrained by the pod-template-hash.
+		syncedRS, err := addHashKeyToRSAndPods(deployment, c, rs, getPodList)
+		if err != nil {
+			return nil, nil, err
+		}
+		syncedRSList = append(syncedRSList, *syncedRS)
+	}
+	syncedPodList, err := getPodList(namespace, options)
+	return syncedRSList, syncedPodList, nil
+}
+
+// addHashKeyToRSAndPods adds pod-template-hash information to the given rs, if it's not already there, with the following steps:
+// 1. Add hash label to the rs's pod template
```

Comment:
Is there a reason you split the steps back to 4? I would expect:

1. update all existing pods
2. update rs podtemplate and selector in one update
3. look for orphaned pods

to be the same plus being cheaper.